tournament.cpp

```diff
@@ -82,58 +87,101 @@ bool CTournament::Init(const CModuleScanner& EngineModule, const CModuleScanner&
 	if(pause)
 		m_DispatchPauseRoute.RouteVirtualFunction(pause, &ConCommand::Dispatch, &CTournament::Pause_Callback, false);
 
+	char* os;
+
+	os = (char*)"Linux";
+
+
 #ifdef _LINUX
-	void *pStartCompetitiveMatch = ServerModule.FindSymbol("_ZN12CTFGameRules21StartCompetitiveMatchEv");
+
+	os = (char*)"Linux";
+
+	void *pStartCompetitiveMatch                                = ServerModule.FindSymbol(
+	"_ZN12CTFGameRules21StartCompetitiveMatchEv");
+	void *CanPlayerChooseClass                                  = ServerModule.FindSymbol(
+	"_ZN12CTFGameRules20CanPlayerChooseClassEP11CBasePlayeri");
+
+	cmd_source                                                  = (int*)EngineModule.FindSymbol(
+	"cmd_source");
+	cmd_clientslot                                              = (int*)EngineModule.FindSymbol(
+	"cmd_clientslot");
+	g_sv_pure_mode                                              = (int*)EngineModule.FindSymbol(
+	"_ZL14g_sv_pure_mode");
+
+	// CanPlayerChooseClass calls another function that calls IsInTournamentMode, so we need the address of that other function
+	// unsigned long CanPlayerChooseClass_TournamentCall       = (unsigned long)((unsigned char*)CanPlayerChooseClass + 0x16);
+	// unsigned long CanPlayerChooseClass_TournamentOffset     = *(unsigned long*)(CanPlayerChooseClass_TournamentCall);
+	// unsigned long CanPlayerChooseClass_Tournament           = CanPlayerChooseClass_TournamentOffset + CanPlayerChooseClass_TournamentCall + 0x10;
+	// PatchAddress((void*)CanPlayerChooseClass_Tournament, 0x18, 1, (unsigned char*)"\xEB");
+	// PatchAddress((void*)CanPlayerChooseClass_Tournament, 0, 1, (unsigned char*)"\xEB");
+	// THIS DOESN'T WORK. Has it ever? I don't think so.
```

Comment:
Yes, this was 100% working, but the offsets most likely changed now.

Reply:
I can't find the offsets in IDA lol. I spent quite a bit of time hunting em down and badgering people in the AlliedMods discord, and they couldn't figure it out either. Regardless, limiting players being unable to choose class thru default TF2's methods is buggy anyway, as you can just click the class repeatedly and bypass it - I'd use https://forums.alliedmods.net/showthread.php?p=642353 for this purpose instead. If you can find the offsets that'd work as well, but it's commented as it doesn't work on current TF2 w/ the current offsets

Reply:
On Linux, `CTFGameRules::CanPlayerChooseClass` calls `CTFGameRules::GetClassLimit`, which calls `CTeamplayRoundBasedRules::IsInTournamentMode`.

Anyway, for Linux; this should just work:
```
void *CTFGameRules_GetClassLimit = ServerModule.FindSymbol("_ZN12CTFGameRules13GetClassLimitEi);
PatchAddress((void*)CTFGameRules_GetClassLimit, 0x18, 1, (unsigned char*)"\xEB");
```

Reply:
I don't have an up to date server.dll, but for Windows it may be:

```
void *CTFGameRules_GetClassLimit = ServerModule.FindSignature(
				(unsigned char*)"\x55\x8B\xEC\x56\x8B\xF1\xE8\x00\x00\x00\x00\x84\xC0\x75\x32\x83","xxxxxxxxxx?");
PatchAddress((void*)CTFGameRules_GetClassLimit, 0xD, 1, (unsigned char*)"\xEB");
```

Reply:
This should be preventing players from switching classes based on the value of tf_tournament_classlimit_<x>, correct? If so, neither of these work, despite seemingly being correct addrs

Reply:
This is to allow the tf_tournament_classlimit_* to work without tournament mode

Reply:
Yeah in that case, no dice